monai/bundle/config_parser.py

```diff
@@ -154,9 +154,25 @@ def __setitem__(self, id: Union[str, int], config: Any):
             self.ref_resolver.reset()
             return
         keys = str(id).split(ID_SEP_KEY)
-        # get the last parent level config item and replace it
-        last_id = ID_SEP_KEY.join(keys[:-1])
-        conf_ = self[last_id]
+        conf_ = self.get()
+        if recursive:
+            if conf_ is None:
+                conf_ = [] if keys[0] == "0" else {}
+            for i, k in enumerate(keys[:-1]):
+                if isinstance(conf_, dict):
+                    if k not in conf_:
+                        conf_[k] = [] if keys[i + 1] == "0" else {}
+                if isinstance(conf_, list):
+                    if int(k) == len(conf_):
```

Comment:
I think this way of appending new elements to a list by index is quite confusing...

Reply:
Hi @wyli ,

Thanks for your review.
I added docstring to describe the feature, it's useful in the data analysis of Auto3D to predefine the report template: https://github.com/Project-MONAI/MONAI/pull/4765#discussion_r955224503, CC @mingxin-zheng .
And another typical use case is to add a transform or handler when overriding a config file.
Refer to Hydra argument: https://github.com/facebookresearch/hydra/blob/v1.2.0/hydra/_internal/instantiate/_instantiate2.py#L155

Thanks.

Reply:
it doesn't make sense in terms of the python style...
it's not ok to set an array with a new index
```py
array = [0, 1]
array[2] = 2  # IndexError: list assignment index out of range
```
while it's ok to set a dictionary with a new key:
```py
d = {"a": 0}
d['b'] = 1
```

Reply:
Hi @wyli ,

Maybe I didn't describe the PR clearly.
The usage is not `array[2] = 2`, which is against python style.
Instead, I just added the `recursive` option to the `set()` function, similar to the enhancement on `get()`, we added support to give the default value in the `get()` function.
So the usage will be:
```py
array = ConfigParser([0, 1])
array.set(id="2") = 2

array[3] = 3  # IndexError
```
What do you think?

Thanks.


Thanks.

Reply:
Simplified to only support nested dicts.

Thanks.